### PR TITLE
Update i3 to 4.21

### PR DESCRIPTION
**Pull request by holiman**
Got a bit further on the pipeline, it seems to mostly build things for me locally, making a PR here to see how it goes for the bot.

### CODE REVIEW

**Comment by holiman** [comment]
When I run the build locally now, it ends with 
```
xplicitly decompress any DWARF compressed ELF sections in /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-input
explicitly decompress any DWARF compressed ELF sections in /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3bar
explicitly decompress any DWARF compressed ELF sections in /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-msg
explicitly decompress any DWARF compressed ELF sections in /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-nagbar
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-config-wizard
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-dump-log
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3bar
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-nagbar
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-input
extracting debug info from /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-msg
original debug info size: 1544kB, size after compression: 1028kB
/usr/lib/rpm/sepdebugcrcfix: Updated 7 CRC32s, 0 CRC32s did match.
3256 blocks
+ /usr/lib/rpm/check-buildroot
Binary file /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-config-wizard matches
Binary file /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3 matches
Found '/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64' in installed files; aborting
error: Bad exit status from /var/tmp/rpm-tmp.s2V66R (%install)

```
I am not sure what the check-buildroot expects

**Comment by holiman** [comment]
> I am not sure what the check-buildroot expects

Ah, so if I understand correctly, the check-buildroot is a script that searches through the binaries, and finds that two of the binaries have the buildroot path baked in, which is a no-no. 

I've tried adding `buildtype=release --strip` to the meson config, but doesn't make any difference.

```
[user@qubes-builder qubes-builder]$ strings chroot-dom0-fc32//home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3-config-wizard | grep BUILDROOT
/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/i3/config.keycodes
/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/xdg
/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/i3/config
[user@qubes-builder qubes-builder]$ strings chroot-dom0-fc32//home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/bin/i3 | grep BUILDROOT
Unable to find the configuration file (looked at $XDG_CONFIG_HOME/i3/config, ~/.i3/config, $XDG_CONFIG_DIRS/i3/config and /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/i3/config)
/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/xdg
/home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64/etc/i3/config
```

https://github.com/i3/i3/blob/next/i3-config-wizard/main.c#L56

EDIT: Yay, now the dom0 build works!

**Comment by holiman** [comment]
Seems to mostly work now. But I don't quite understand what is meant to be included in the non-dom0 packages. Are those supposed to be pure meta-packages or should they also include the actual `i3` binaries? 

As it is now, in this PR, they do actually build the i3 sources, but I'm wondering if that's just some side-effect of `configure` triggering a `meson` build or something. 

I guess the question is: What does the non-root qubes actually need, in terms of i3 integration?

File: debian-pkg/debian/changelog [hunk 1]
```diff
@@ -1,3 +1,10 @@
+i3-wm (4.21-1) unstable; urgency=medium
+
+  [ Martin Holst Swende (holiman) ]
```
Comment (holiman) [inline debian-pkg/debian/changelog:3]: Should add @anadahz too

File: i3.spec.in [hunk 2]
```diff
@@ -124,27 +125,36 @@ find . -maxdepth 1 -type f -name "i3*" -exec sed -i -e '1s;^#!/usr/bin/env perl;
...
+# See https://mesonbuild.com/Builtin-options.html for Meson built-in options
+meson setup --prefix=/usr build && pushd build
+meson compile
+
+# echo "buildroot: %{buildroot} _bindir: %{_bindir} datadir : %{_datadir} name %{name}"
+#  buildroot: /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64 _bindir: /usr/bin datadir : /usr/share name i3
+
 popd
 
+echo "TODO: generate doxygen"
```
Comment (holiman) [inline i3.spec.in:136]:
```suggestion
```

File: i3.spec.in [hunk 3]
```diff
@@ -124,27 +125,36 @@ find . -maxdepth 1 -type f -name "i3*" -exec sed -i -e '1s;^#!/usr/bin/env perl;
...
-ln -s ../configure configure
-%configure
-%make_build
+
+# See https://mesonbuild.com/Builtin-options.html for Meson built-in options
+meson setup --prefix=/usr build && pushd build
+meson compile
+
+# echo "buildroot: %{buildroot} _bindir: %{_bindir} datadir : %{_datadir} name %{name}"
+#  buildroot: /home/user/rpmbuild/BUILDROOT/i3-4.21-6.fc32.x86_64 _bindir: /usr/bin datadir : /usr/share name i3
```
Comment (holiman) [inline i3.spec.in]:
```suggestion
```

**Comment by holiman** [comment]
I've been testing this version in dom0 for a few days now. I was somewhat hoping that it would solve my occasional crashes [reported here](https://github.com/QubesOS/qubes-issues/issues/7782), but that does not seem to be the case. 

I don't really notice any change at all, to be honest. LMK if there's anything in particular I should test.

**Comment by holiman** [comment]
Been running this for a couple of weeks now, seems to work fine. Is there anything that I can do to help move this along?

**Review by marmarek — APPROVED** [review]

**Comment by marmarek** [comment]
Thanks!

**Comment by Minimalist73** [comment]
> LMK if there's anything in particular I should test.

A bit late sorry, but can you try to reload i3 and see if your widgets disappear (sys-net network manager for example)?

**Comment by holiman** [comment]
Neither reload `meta-shift-c` nor inpace-restart `meta-shfit-r` causes any disappearing widgets. A brief screen flicker, then everything back as it was.

**Comment by Minimalist73** [comment]
> Neither reload `meta-shift-c` nor inpace-restart `meta-shfit-r` causes any disappearing widgets. A brief screen flicker, then everything back as it was.

Well that's strange. I just tested using the new version and my widgets (coming from qubes, not dom0) are gone after both reload and restart. Maybe it's because of my current bar, what do you use and what are your options with it?

**Comment by holiman** [comment]
I think mine is pretty vanilla, just `status_command qubes-i3status`, it just uses the default `i3bar`.

**Comment by networkjanitor** [comment]
Can reproduce disappearing tray widgets (nm-applet, whonix) with vanilla config and both `meta-shift-c` and `meta-shift-r`.

However after some testing they seem to either completely disappear (no remaining visible window on any workspace/tray) or just detach from tray and become their own window.

**Pull request merged**